File: src/main/java/org/dukcode/ps/codetree/trail02/chapter03/lesson03/ChallengeDistanceFromOrigin.java
package org.dukcode.ps.codetree.trail02.chapter03.lesson03;

import java.io.BufferedReader;
import java.io.BufferedWriter;
import java.io.IOException;
import java.io.InputStreamReader;
import java.io.OutputStreamWriter;
import java.util.Arrays;
import java.util.Comparator;
import java.util.StringTokenizer;

/**
 * @see <a
 * href="https://www.codetree.ai/ko/trails/complete/curated-cards/challenge-distance-from-origin/description">LINK</a>
 */
public class ChallengeDistanceFromOrigin {

  private static BufferedReader br;
  private static BufferedWriter bw;
  private static StringTokenizer st;

  public static void main(String[] args) throws IOException {
    br = new BufferedReader(new InputStreamReader(System.in));
    bw = new BufferedWriter(new OutputStreamWriter(System.out));

    int n = Integer.parseInt(br.readLine());
    Position[] positions = new Position[n];
    for (int idx = 0; idx < n; idx++) {
      st = new StringTokenizer(br.readLine());
      int x = Integer.parseInt(st.nextToken());
      int y = Integer.parseInt(st.nextToken());

      positions[idx] = new Position(idx, x, y);
    }

    Arrays.sort(positions,
        Comparator.comparingInt(Position::getDistance).thenComparing(p -> p.idx));

    for (Position position : positions) {
      bw.write(String.valueOf(position.idx + 1));
      bw.newLine();
    }
    br.close();
    bw.close();

  }

  private static class Position {

    int idx;
    int x;
    int y;

    public Position(int idx, int x, int y) {
      this.idx = idx;
      this.x = x;
      this.y = y;
    }

    public int getDistance() {
      return Math.abs(this.x) + Math.abs(this.y);
    }


  }
}

File: src/main/kotlin/org/dukcode/ps/codetree/trail02/chapter03/lesson03/ChallengeDistanceFromOrigin.kt
package org.dukcode.ps.codetree.trail02.chapter03.lesson03.challengeDistanceFromOrigin

import java.io.BufferedReader
import java.io.BufferedWriter
import java.io.InputStreamReader
import java.io.OutputStreamWriter
import java.util.StringTokenizer
import kotlin.math.abs

private val br = BufferedReader(InputStreamReader(System.`in`))
private val bw = BufferedWriter(OutputStreamWriter(System.out))

/**
 * @see <a href="https://www.codetree.ai/ko/trails/complete/curated-cards/challenge-distance-from-origin/description">LINK</a>
 */
fun main() {
    val n = br.readLine().toInt()
    val points =
        List(n) {
            val (x, y) = parseInts()
            Point(it + 1, x, y)
        }.sortedWith(compareBy(Point::distance, Point::number))

    points.forEach {
        bw.write(it.number.toString())
        bw.newLine()
    }

    bw.flush()
}

private class Point(
    val number: Int,
    val x: Int,
    val y: Int,
) {
    val distance: Int
        get() = abs(x) + abs(y)
}

private fun parseInts() =
    StringTokenizer(br.readLine()).run {
        IntArray(countTokens()) {
            nextToken().toInt()
        }
    }
